src/XMLInterface.cpp

```diff
@@ -519,3 +521,108 @@ void XMLInterface::writeXML(Scores& fullset, ProteinProbEstimator* protEstimator
   os.close();
 }
 
+void XMLInterface::writePeptideXML(Scores& fullset, ProteinProbEstimator* protEstimator, std::string call) {
+
+  
+  ofstream os;
+  const string schema = // space +
+      "http://sashimi.sourceforge.net/schema_revision/pepXML/pepXML_v122.xsd";
+
+
+  os.open(xmlPeptideOutputFN_.data(), ios::out | ios::binary);
+  os << "<?xml version=\"1.0\" encoding=\"UTF-8\"?>" << endl;
+  os << "<root xmlns:ns0=\"http://regis-web.systemsbiology.net/pepXML\">\n"<< endl;
+
+  std::string pepPath = xmlPeptideOutputFN_.data();
+  /* TODO: MUST FIX */
+  std::string base_name = pepPath.substr(0, pepPath.find("."));
+
+  ifstream ifs_psms(xmlpeptideOutputFN_PSMs.data(), ios::in | ios::binary);
+  os << ifs_psms.rdbuf();
+  ifs_psms.close();
+  remove(xmlpeptideOutputFN_PSMs.c_str());
+
+
+  os << "    </ns0:msms_run_summary>" << endl;
+  
+  os << "</root>" << endl;
+  os.close();
+}
+
+// Change this function name to something including a string pepXML
+void XMLInterface::writePeptideXML_PSMs(Scores& fullset, double selectionFdr_) {
+
+  pi0Psms_ = fullset.getPi0();
+  numberQpsms_ = fullset.getQvaluesBelowLevel(0.01);
+
+  ofstream os;
+  
+  /* xmlpeptideOutputFN_ = xmlPeptideOutputFN_; */
+  xmlpeptideOutputFN_PSMs.append("writePeptideXML_PSMs");
+  os.open(xmlpeptideOutputFN_PSMs.c_str(), ios::out);
+
+  /* os << "  <psms>" << endl; */
+  map<char,float> aaDict = getRoughAminoWeightDict();
+  /* Sort psms based on base name  */
+  std::sort(fullset.begin(), fullset.end(), less_than_base_name());
+  std::string pepXMLBaseName = "";
+  bool first_msms_summary = true;
+
+  for (std::vector<ScoreHolder>::iterator sh = fullset.begin();sh != fullset.end(); ++sh) {
+    std::string id = sh->pPSM->getId();
+    auto baseName = id.substr(0, id.find('.'));
+    if (baseName != pepXMLBaseName) {
+      /*  Start of a new msms run*/
+      pepXMLBaseName = baseName;
+      if (first_msms_summary) {
+        first_msms_summary = false;
+      } else {
+        /* End of msms run */
+        os << "    </ns0:msms_run_summary>" << endl;
+      }
+    
+      /* New msms run! */
+      os << "    <ns0:msms_run_summary base_name=\"" << baseName << "\">" << endl;
+      os << "    <ns0:search_summary>" << endl;
+      os << "    <ns0:parameter name=\"decoy_prefix\" value=\"rev_\" />" << endl;
+      os << "    </ns0:search_summary>" << endl;
+    }
+    if (sh->q < selectionFdr_)
+      sh->printPepXML(os, aaDict);
+  }
+  /* os << "  </psms>" << endl << endl; */
+  os.close();
+}
+
+
+
+map<char, float> XMLInterface::getRoughAminoWeightDict() {
+  // This function gives an approximate amino acid weight
+  // with two digits precision. For parsing purposes.
+  map<char, float> roughAminoAcidWeight =
```

Comment:
Could we store this array in a static, so that it does not have to be recreated each time the function is called? E.g. a [singleton pattern](https://en.wikipedia.org/wiki/Singleton_pattern)